.cursor/rules/api-specific-config.mdc

```diff
@@ -21,17 +21,26 @@ alwaysApply: false
   - Use `DISTINCT ON` with proper indexes for latest-record queries
   - Use database-level `GROUP BY` for grouping operations
   - Use CTEs (Common Table Expressions) for complex multi-step queries
+  - Use `ORDER BY` in SQL instead of JavaScript `.sort()`
+  - For complex sorts (like `DISTINCT ON` + different `ORDER BY`), use subqueries with `.as()` in Drizzle
   - Example: Getting latest summaries should use `DISTINCT ON (agent_id)` not fetch all and filter in memory
+- **Response Size Management**:
+  - Never return unbounded result sets for in-memory processing
+  - Always aggregate at database level when only totals are needed
+  - Example: Use `SUM(total_volume)` in SQL vs fetching 1000+ records to sum in JavaScript
+  - Set reasonable LIMIT clauses for safety even on internal queries
+  - Monitor query response sizes in production
 - **Composite Indexes**: Always add indexes for:
   - Foreign key relationships
   - Columns used in `WHERE` clauses
   - Columns used in `ORDER BY` with `DISTINCT ON`
   - Multi-column patterns that match query patterns (e.g., `(agent_id, competition_id, timestamp DESC)`)
 - **Query Optimization Principles**:
-  - Avoid N+1 queries - use joins or batch fetching
+  - Prevent the N+1 query problem: Instead of fetching a list then making separate queries per item, use joins or batch fetching
```

Comment:
```suggestion
  - Prevent the linear query scaling problem: Instead of fetching a list then making separate queries per item, use joins or batch fetching
```